src/mailadm/conn.py

```diff
@@ -210,31 +185,36 @@ def get_user_list(self, token=None):
             args.append(token)
         return [UserInfo(*args) for args in self._sqlconn.execute(q, args).fetchall()]
 
-    def add_email_account(self, token_info, addr=None, password=None, tries=1):
-        for i in range(tries):
-            try:
-                return self._add_addr(token_info, addr=addr, password=password)
-            except (ValueError, DBError):
-                if i + 1 >= tries:
-                    raise
-
-    def _add_addr(self, token_info, addr, password):
-        config = self.config
+    def add_email_account(self, token_info, addr=None, password=None):
+        token_info.check_exhausted()
+        if password is None:
+            password = mailadm.util.gen_password()
         if addr is None:
             rand_part = mailadm.util.get_human_readable_id()
             username = "{}{}".format(token_info.prefix, rand_part)
-            addr = "{}@{}".format(username, config.mail_domain)
+            addr = "{}@{}".format(username, self.config.mail_domain)
         else:
-            if not addr.endswith(config.mail_domain):
+            if not addr.endswith(self.config.mail_domain):
                 raise ValueError("email {!r} is not on domain {!r}".format(
-                                 addr, config.mail_domain))
+                                 addr, self.config.mail_domain))
 
-        clear_pw, hash_pw = get_doveadm_pw(password=password)
-        self.add_user(addr=addr, hash_pw=hash_pw, date=int(time.time()),
+        self.add_user(addr=addr, date=int(time.time()),
                       ttl=token_info.get_expiry_seconds(), token_name=token_info.name)
-        user_info = self.get_user_by_addr(addr)
+
         self.log("added addr {!r} with token {!r}".format(addr, token_info.name))
-        user_info.clear_pw = clear_pw
+
+        user_info = self.get_user_by_addr(addr)
+        user_info.password = password
+
+        # seems that everything is fine so far, so let's invoke mailcow:
+        mailcow = MailcowConnection(self.config)
+        try:
+            mailcow.add_user_mailcow(addr, password)
+        except MailcowError:
+            self.del_user(addr)
+            self.mod_token(token_info.name, maxuse=token_info.maxuse + 1)
+            raise
```

Comment:
why is this neccessary?  The cmdline should call "db.rollback()" in case of exceptions -- or rather the "write_transaction()" context manager does.  Or am i missing something about this maxuse-incrementing logic?  It seems weird to increment maxuse by one anyway.

Reply:
That's super interesting. I changed it to do a self.rollback().

but - shouldn't it do the rollback automatically when there is an exception, according to https://github.com/deltachat/mailadm/pull/25/files#diff-6582a0223c7ac9e73fefd6d9f4c9ceffaac128c554c68ecfc999db8b402a7f54R58 ? Somehow this doesn't work. I guess it has something to do with how the connection is handled in cmdline.py. Any ideas?

Reply:
I'm not *exactly* sure why it works now, but it works now.